core/src/main/java/hudson/tasks/BuildStepCompatibilityLayer.java

```diff
@@ -70,16 +70,20 @@ public boolean prebuild(AbstractBuild<?,?> build, BuildListener listener) {
     @Override
     public boolean perform(AbstractBuild<?,?> build, Launcher launcher, BuildListener listener) throws InterruptedException, IOException {
         if (this instanceof SimpleBuildStep) {
+            final SimpleBuildStep sbs = ((SimpleBuildStep) this);
             // delegate to the overloaded version defined in SimpleBuildStep
             FilePath workspace = build.getWorkspace();
-            if (workspace == null) {
+            if (workspace == null && sbs.requiresWorkspace()) {
                 throw new AbortException("no workspace for " + build);
             }
-            ((SimpleBuildStep) this).perform(build, workspace, build.getEnvironment(listener), launcher, listener);
+            if (launcher == null && sbs.requiresLauncher()) {
+                throw new AbortException("no launcher for " + build);
```

Comment:
Similarly, for an `AbstractBuild` this would be a pathological condition. There is probably no real need to patch this class at all.

Reply:
File changes reverted.